File: problem003/euler3.java
public class euler3 {

	public static void main(String[] args){
		long n = 600851475143L;
		int i = 2;
		System.out.println(factor(n, i));
	}

	private static long factor(long n, int i){
		long maxfactor = 2;
		while (i < n/2){
			if (n % i == 0){
				n = n / i;
				maxfactor = n;
			} else { i += 1; }
		}
		return maxfactor;
	}
}

File: problem003/euler3.scala
object euler3 {

    def factor(i: Int, n: Long, maxFactor: Long) : Long = {
        if (i < n / 2) {
            if (n % i == 0)
                factor(i, n / i, n / i)
            else
                factor(i + 1, n, maxFactor)
        } else
            maxFactor
    }

    def main(args: Array[String]) {
        val n = 600851475143L;
        println(factor(2, n, n))
    }
}
